docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,720 @@
+
+.. note::
+
+    Note that ``read_global_safe`` that used to take an ``uint32_t`` as the index type, takes a signed integer now. When indexing an array with unsigned integers, the compiler has to handle integer overflows, as they're defined by the C/C++ standards. It might happen, that some part of the vector load indices overflow, thus resulting in a non-contiguous read. If you change the previously linked code to use an unsigned integer as the thread id, the compiler won't emit a vector load. Signed integer overflow is an undefined behavior, and hence unknown to the optimizer. To convey the absence of overflow to the compiler with unsigned indices, add ``__builtin_assume(gid + 4 > gid)``, or the more portable ``[[assume]](gid + 4 > gid)``, once ``amdclang++`` supports it.
+
+``read_global_safe`` implementation is an Immediately Invoked Lambda Expression (IILE), because ``ItemsPerThread`` is an integer value, while you need a compile-time ``iota``-like sequence of integers as a pack for the fold-expressions to expand on. This can only take place as part of template argument deduction on the IILE.
+
+.. _processing-items:
+
+Processing ``ItemsPerThread``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once the kernel reads ``ItemsPerThread`` number of inputs to local, it immediately reduces them to a scalar. There is no reason to propagate the input element multiplicity to the warp reduction phase.
+
+.. code-block:: C++
+
+    T res = [&]()
+    {
+        // Read input from front buffer to local
+        hip::static_array<T, ItemsPerThread> arr = read_global_safe(gid);
+
+        // Reduce ItemsPerThread to scalar
+        tmp::static_for<1, tmp::less_than<ItemsPerThread>, tmp::increment<1>>([&]<int I>()
+        {
+            get<0>(arr) = op(get<0>(arr), get<I>(arr));
+        });
+
+        return get<0>(arr);
+    }();
+
+Two-pass reduction
+------------------
+
+Alter kernel launch and input fetching such that no more blocks are launched than what a subsequent kernel launch's single block can conveniently reduce, while performing multiple passes of input reading from global and combining their results before engaging in the end game tree-like reduction.
+
+With this method, you can save at least one to two kernel launches for really large inputs.
+
+Global data share
+-----------------
+
+.. warning::
+
+    This modification can only be executed on AMD hardware.
+
+Perform the first step of the two-pass reduction, but in the end, instead of writing to global and reading it back in a subsequent kernel, write the partial results to the Global Data Share (GDS). This is an ``N+1`` th shared memory that is accessed by all multiprocessors and is also on-chip memory.
```

Comment:
```suggestion
Perform the first step of the two-pass reduction, but in the end, instead of writing to global and reading it back in a subsequent kernel, write the partial results to the Global Data Share (GDS). This is an ``N+1`` **th** shared memory that is accessed by all multiprocessors and is also on-chip memory.
```

Reply:
what is the th?

Reply:
I would not make it bold.  The problem: https://www.reddit.com/r/math/comments/nw5lb1/n1st_or_n1th/